runtime/cudaq/platform/default/rest_server/RemoteSimulatorQPU.cpp

```diff
@@ -0,0 +1,357 @@
+        std::string errMsg;
+        llvm::raw_string_ostream os(errMsg);
+        const std::string pipeline =
+            std::accumulate(clientPasses.begin(), clientPasses.end(),
+                            std::string(), [](const auto &ss, const auto &s) {
+                              return ss.empty() ? s : ss + "," + s;
+                            });
+        if (failed(parsePassPipeline(pipeline, pm, os)))
+          throw std::runtime_error(
+              "Remote rest platform failed to add passes to pipeline (" +
+              errMsg + ").");
+
+        if (failed(pm.run(moduleOp)))
+          throw std::runtime_error(
+              "Remote rest platform: applying IR passes failed.");
+      }
+      std::string mlirCode;
+      llvm::raw_string_ostream outStr(mlirCode);
+      mlir::OpPrintingFlags opf;
+      opf.enableDebugInfo(/*enable=*/true,
+                          /*pretty=*/false);
+      moduleOp.print(outStr, opf);
+      request.code = llvm::encodeBase64(mlirCode);
+    }
+
+    request.simulator = m_simName;
+    request.seed = cudaq::get_random_seed();
+    // Don't let curl adding "Expect: 100-continue" header, which is not
+    // suitable for large requests, e.g., bitcode in the JSON request.
+    //  Ref: https://gms.tf/when-curl-sends-100-continue.html
+    std::map<std::string, std::string> headers{
+        {"Expect:", ""}, {"Content-type", "application/json"}};
+    json requestJson = request;
+    auto resultJs = m_client.post(m_url, "job", requestJson, headers, false);
+    resultJs.get_to(*executionContext);
+  }
+
+  void setExecutionContext(cudaq::ExecutionContext *context) override {
+    cudaq::info("RemoteSimulatorQPU::setExecutionContext QPU {}", qpu_id);
+    std::scoped_lock<std::mutex> lock(m_contextMutex);
+    m_contexts[std::this_thread::get_id()] = context;
+  }
+
+  void resetExecutionContext() override {
+    cudaq::info("RemoteSimulatorQPU::resetExecutionContext QPU {}", qpu_id);
+    std::scoped_lock<std::mutex> lock(m_contextMutex);
+    m_contexts.erase(std::this_thread::get_id());
+  }
+};
+
+// Util to query an available TCP/IP port for auto-launching a server instance.
+std::optional<std::string> getAvailablePort() {
+  int sock = ::socket(AF_INET, SOCK_STREAM, 0);
+  if (sock < 0)
+    return {};
+  struct sockaddr_in servAddr;
+  ::bzero((char *)&servAddr, sizeof(servAddr));
+  servAddr.sin_family = AF_INET;
+  servAddr.sin_addr.s_addr = INADDR_ANY;
+  // sin_port = 0 => auto assign
+  servAddr.sin_port = 0;
+  if (::bind(sock, (struct sockaddr *)&servAddr, sizeof(servAddr)) < 0)
+    return {};
+  socklen_t len = sizeof(servAddr);
+  if (::getsockname(sock, (struct sockaddr *)&servAddr, &len) == -1)
+    return {};
+  if (close(sock) < 0)
+    return {};
+  return std::to_string(::ntohs(servAddr.sin_port));
+}
+
+// Multi-QPU platform for remotely-hosted simulator QPUs
+class RemoteSimulatorQuantumPlatform : public cudaq::quantum_platform {
```

Comment:
Can this be merged with the existing MultiQPUPlatform? (comment / question from today's walk through)